File: android/src/main/java/com/unact/yandexmapkit/full/UtilsFull.java
package com.unact.yandexmapkit.full;

import com.unact.yandexmapkit.lite.UtilsLite;
import com.yandex.mapkit.RequestPoint;
import com.yandex.mapkit.RequestPointType;
import com.yandex.mapkit.annotations.AnnotationLanguage;
import com.yandex.mapkit.directions.driving.AvoidanceFlags;
import com.yandex.mapkit.directions.driving.DrivingOptions;
import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.search.SearchOptions;
import com.yandex.mapkit.search.SuggestOptions;
import com.yandex.mapkit.transport.masstransit.FitnessOptions;
import com.yandex.mapkit.transport.masstransit.TimeOptions;

import java.util.Map;

public class UtilsFull extends UtilsLite {
  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public static RequestPoint requestPointFromJson(Map<String, Object> json) {
    return new RequestPoint(
      pointFromJson((Map<String, Object>) json.get("point")),
      RequestPointType.values()[(Integer) json.get("requestPointType")],
      (String) json.get("pointContext"),
      (String) json.get("drivingArrivalPointId"),
      (String) json.get("indoorLevelId")
    );
  }
  public static TimeOptions timeOptionsFromJson(Map<String, Object> json) {
    return new TimeOptions(
      (Long) json.get("departureTime"),
      (Long) json.get("arrivalTime")
    );
  }

  @SuppressWarnings({"ConstantConditions"})
  public static FitnessOptions fitnessOptionsFromJson(Map<String, Object> json) {
    return new FitnessOptions(
      (Boolean) json.get("avoidSteep"),
      (Boolean) json.get("avoidStairs")
    );
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public static DrivingOptions drivingOptionsFromJson(Map<String, Object> json) {
    AvoidanceFlags avoidanceFlags = json.get("avoidanceFlags") != null ?
      avoidanceFlagsFromJson((Map<String, Object>) json.get("avoidanceFlags")) :
      null;
    AnnotationLanguage annotationLanguage = json.get("annotationLanguage") != null ?
      AnnotationLanguage.values()[(Integer) json.get("annotationLanguage")] :
      null;

    return new DrivingOptions(
      (Double) json.get("initialAzimuth"),
      (Integer) json.get("routesCount"),
      (Long) json.get("departureTime"),
      annotationLanguage,
      avoidanceFlags
    );
  }

  @SuppressWarnings({"ConstantConditions"})
  public static AvoidanceFlags avoidanceFlagsFromJson(Map<String, Object> json) {
    return new AvoidanceFlags(
      (Boolean) json.get("avoidTolls"),
      (Boolean) json.get("avoidUnpaved"),
      (Boolean) json.get("avoidPoorCondition"),
      (Boolean) json.get("avoidRailwayCrossing"),
      (Boolean) json.get("avoidBoatFerry"),
      (Boolean) json.get("avoidFordCrossing"),
      (Boolean) json.get("avoidTunnel"),
      (Boolean) json.get("avoidHighway")
    );
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public static SearchOptions searchOptionsFromJson(Map<String, Object> json) {
    Point userPosition = json.get("userPosition") != null ?
      pointFromJson((Map<String, Object>) json.get("userPosition")) :
      null;

    return new SearchOptions(
      ((Number) json.get("searchType")).intValue(),
      (Integer) json.get("resultPageSize"),
      0,
      userPosition,
      (String) json.get("origin"),
      (Boolean) json.get("geometry"),
      (Boolean) json.get("disableSpellingCorrection"),
      null
    );
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public static SuggestOptions suggestOptionsFromJson(Map<String, Object> json) {
    Point userPosition = json.get("userPosition") != null ?
      pointFromJson((Map<String, Object>) json.get("userPosition")) :
      null;

    return new SuggestOptions(
      ((Number) json.get("suggestType")).intValue(),
      userPosition,
      ((Boolean) json.get("suggestWords")),
      ((Boolean) json.get("strictBounds"))
    );
  }
}

File: ios/Classes/full/UtilsFull.swift
import YandexMapsMobile

class UtilsFull: UtilsLite {
  static func requestPointFromJson(_ json: [String: Any]) -> YMKRequestPoint {
    let point = pointFromJson(json["point"] as! [String: NSNumber])
    let pointType = YMKRequestPointType(rawValue: (json["requestPointType"] as! NSNumber).uintValue)!

    return YMKRequestPoint(
      point: point,
      type: pointType,
      pointContext: json["pointContext"] as? String,
      drivingArrivalPointId: json["drivingArrivalPointId"] as? String,
      indoorLevelId: json["indoorLevelId"] as? String
    )
  }

  static func timeOptionsFromJson(_ json: [String: Any]) -> YMKTimeOptions {
    return YMKTimeOptions(
      departureTime: json["departureTime"] as? NSNumber == nil ?
        nil :
        Date(timeIntervalSince1970: (json["departureTime"] as! NSNumber).doubleValue / 1000.0),
      arrivalTime: json["arrivalTime"] as? NSNumber == nil ?
        nil :
        Date(timeIntervalSince1970: (json["arrivalTime"] as! NSNumber).doubleValue / 1000.0)
    )
  }

  static func fitnessOptionsFromJson(_ json: [String: Any]) -> YMKFitnessOptions {
    return YMKFitnessOptions(
      avoidSteep: (json["avoidSteep"] as! NSNumber).boolValue,
      avoidStairs: (json["avoidStairs"] as! NSNumber).boolValue
    )
  }

  static func drivingOptionsFromJson(_ json: [String: Any]) -> YMKDrivingOptions {
    return YMKDrivingOptions(
      initialAzimuth: json["initialAzimuth"] as? NSNumber,
      routesCount: json["routesCount"] as? NSNumber,
      departureTime: json["departureTime"] as? NSNumber == nil ?
        nil :
        Date(timeIntervalSince1970: (json["departureTime"] as! NSNumber).doubleValue / 1000.0),
      annotationLanguage: json["annotationLanguage"] as? NSNumber,
      avoidanceFlags: json["avoidanceFlags"] as? [String: Any] == nil ?
        nil :
        avoidanceFlagsFromJson(json["avoidanceFlags"] as! [String: Any])
    )
  }

  static func avoidanceFlagsFromJson(_ json: [String: Any]) -> YMKDrivingAvoidanceFlags {
    return YMKDrivingAvoidanceFlags(
      avoidTolls: (json["avoidTolls"] as! NSNumber).boolValue,
      avoidUnpaved: (json["avoidUnpaved"] as! NSNumber).boolValue,
      avoidPoorCondition: (json["avoidPoorCondition"] as! NSNumber).boolValue,
      avoidRailwayCrossing: (json["avoidRailwayCrossing"] as! NSNumber).boolValue,
      avoidBoatFerry: (json["avoidBoatFerry"] as! NSNumber).boolValue,
      avoidFordCrossing: (json["avoidFordCrossing"] as! NSNumber).boolValue,
      avoidTunnel: (json["avoidTunnel"] as! NSNumber).boolValue,
      avoidHighway: (json["avoidHighway"] as! NSNumber).boolValue
    )
  }

  static func searchOptionsFromJson(_ json: [String: Any]) -> YMKSearchOptions {
    let userPosition = json["userPosition"] as? [String: Any] != nil ?
      pointFromJson(json["userPosition"] as! [String: NSNumber]) :
      nil

    return YMKSearchOptions(
      searchTypes: YMKSearchType(rawValue: (json["searchType"] as! NSNumber).uintValue),
      resultPageSize: json["resultPageSize"] as? NSNumber,
      snippets: YMKSearchSnippet(),
      userPosition: userPosition,
      origin: json["origin"] as? String,
      geometry: (json["geometry"] as! NSNumber).boolValue,
      disableSpellingCorrection: (json["disableSpellingCorrection"] as! NSNumber).boolValue,
      filters: nil
    )
  }

  static func suggestOptionsFromJson(_ json: [String: Any]) -> YMKSuggestOptions {
    let userPosition = json["userPosition"] as? [String: Any] != nil ?
      pointFromJson(json["userPosition"] as! [String: NSNumber]) :
      nil

    return YMKSuggestOptions(
      suggestTypes: YMKSuggestType.init(rawValue: (json["suggestType"] as! NSNumber).uintValue),
      userPosition: userPosition,
      suggestWords: (json["suggestWords"] as! NSNumber).boolValue,
      strictBounds: (json["strictBounds"] as! NSNumber).boolValue
    )
  }
}
